src/vs/editor/common/model.ts

```diff
@@ -851,7 +851,7 @@ export interface ITextModel {
 	 * Set the current language mode associated with the model.
 	 * @internal
 	 */
-	setMode(languageId: string): void;
+	setMode(languageId: string, reason?: string): void;
```

Comment:
I am not sure if `reason` or better `source`. If we model this after the `setPosition` method on the code editor then it should be `reason` that can also be undefined or null: 

https://github.com/microsoft/vscode/blob/c4722f1bb292284e0437eb7547808a5788fe6430/src/vs/editor/common/viewModel/viewModelImpl.ts#L997

But I refer to code editor reviewers to chime in.

Reply:
funny enough, I had `source` first but swapped it to `reason` :) yes let's see what @hediet has to say.

Reply:
Can you introduce a pseudo type for `reason`, maybe with an associated namespace for well-known reasons?
E.g. like this:
```ts
export type SetModeReason = string;
namespace SetModeReason {
    export const unknown = 'unknown';
    export const languageDetection = 'languageDetection';
}
```

Otherwise it will be wild west and hard to understand its purpose.

Reply:
synced with @hediet on this and we won't go this route to keep things consistent